subprojects/core/src/main/java/org/gradle/api/internal/changedetection/rules/InputPropertiesTaskStateChanges.java

```diff
@@ -35,36 +33,25 @@
     private final Set<String> changed;
     private final Set<String> added;
 
-    public InputPropertiesTaskStateChanges(@Nullable TaskExecution previousExecution, TaskExecution currentExecution, TaskInternal task, ValueSnapshotter valueSnapshotter) {
+    public InputPropertiesTaskStateChanges(@Nullable TaskExecution previousExecution, TaskExecution currentExecution, TaskInternal task) {
         ImmutableSortedMap<String, ValueSnapshot> previousInputProperties = previousExecution == null ? ImmutableSortedMap.<String, ValueSnapshot>of() : previousExecution.getInputProperties();
-        ImmutableSortedMap.Builder<String, ValueSnapshot> builder = ImmutableSortedMap.naturalOrder();
         removed = new HashSet<String>(previousInputProperties.keySet());
         changed = new HashSet<String>();
         added = new HashSet<String>();
-        for (Map.Entry<String, Object> entry : task.getInputs().getProperties().entrySet()) {
+        ImmutableSortedMap<String, ValueSnapshot> currentInputProperties = currentExecution.getInputProperties();
+        for (Map.Entry<String, ValueSnapshot> entry : currentInputProperties.entrySet()) {
             String propertyName = entry.getKey();
-            Object value = entry.getValue();
-            try {
-                removed.remove(propertyName);
-                ValueSnapshot previousSnapshot = previousInputProperties.get(propertyName);
-                if (previousSnapshot == null) {
-                    added.add(propertyName);
-                    builder.put(propertyName, valueSnapshotter.snapshot(value));
-                } else {
-                    ValueSnapshot newSnapshot = valueSnapshotter.snapshot(value, previousSnapshot);
-                    if (newSnapshot == previousSnapshot) {
-                        builder.put(propertyName, previousSnapshot);
-                    } else {
-                        changed.add(propertyName);
-                        builder.put(propertyName, valueSnapshotter.snapshot(value));
-                    }
+            ValueSnapshot currentSnapshot = entry.getValue();
+            removed.remove(propertyName);
+            ValueSnapshot previousSnapshot = previousInputProperties.get(propertyName);
+            if (previousSnapshot == null) {
+                added.add(propertyName);
+            } else {
+                if (!currentSnapshot.equals(previousSnapshot)) {
```

Comment:
That snapshot reuse convention is an implementation detail of `ValueSnapshotter`, and thus we shouldn't rely on that.

This way end up calling `Object.equals()` here, which does exactly an `==`, but with the added benefit that it would keep working even if we modified the implementation of `ValueSnapshotter`.